src/pool/pool_jemalloc.c

```diff
@@ -0,0 +1,345 @@
+                                  unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)addr;         // unused
+    (void)size;         // unused
+    (void)offset;       // unused
+    (void)length;       // unused
+    (void)arena_ind;    // unused
+
+    return true;
+}
+
+static bool arena_extent_purge_lazy(extent_hooks_t *extent_hooks, void *addr,
+                                    size_t size, size_t offset, size_t length,
+                                    unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)size;         // unused
+
+    jemalloc_memory_pool_t *pool = get_kind_by_arena(arena_ind);
+    enum umf_result_t ret = umfMemoryProviderPurgeLazy(
+        pool->hProvider, (char *)addr + offset, length);
+
+    return (ret == UMF_RESULT_SUCCESS);
+}
+
+static bool arena_extent_purge_forced(extent_hooks_t *extent_hooks, void *addr,
+                                      size_t size, size_t offset, size_t length,
+                                      unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)size;         // unused
+
+    jemalloc_memory_pool_t *pool = get_kind_by_arena(arena_ind);
+    enum umf_result_t ret = umfMemoryProviderPurgeForce(
+        pool->hProvider, (char *)addr + offset, length);
+
+    return (ret == UMF_RESULT_SUCCESS);
+}
+
+static bool arena_extent_split(extent_hooks_t *extent_hooks, void *addr,
+                               size_t size, size_t size_a, size_t size_b,
+                               bool committed, unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)addr;         // unused
+    (void)size;         // unused
+    (void)size_a;       // unused
+    (void)size_b;       // unused
+    (void)committed;    // unused
+    (void)arena_ind;    // unused
+
+    return false;
+}
+
+static bool arena_extent_merge(extent_hooks_t *extent_hooks, void *addr_a,
+                               size_t size_a, void *addr_b, size_t size_b,
+                               bool committed, unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)addr_a;       // unused
+    (void)addr_b;       // unused
+    (void)size_a;       // unused
+    (void)size_b;       // unused
+    (void)committed;    // unused
+    (void)arena_ind;    // unused
+
+    return false;
+}
+
+static extent_hooks_t arena_extent_hooks = {
+    .alloc = arena_extent_alloc,
+    .dalloc = arena_extent_dalloc,
+    .destroy = arena_extent_destroy,
+    .commit = arena_extent_commit,
+    .decommit = arena_extent_decommit,
+    .purge_lazy = arena_extent_purge_lazy,
+    .purge_forced = arena_extent_purge_forced,
+    .split = arena_extent_split,
+    .merge = arena_extent_merge};
+
+static void *je_malloc(void *pool, size_t size) {
+    jemalloc_memory_pool_t *je_pool = (jemalloc_memory_pool_t *)pool;
+    unsigned arena = je_pool->arena_zero;
+    int flags = MALLOCX_ARENA(arena) | MALLOCX_TCACHE_NONE;
```

Comment:
Please add a comment why we OR with TCACHE_NONE

Reply:
because it's a bit flag? i'm not sure what you are asking.

The reason tcache needs to be set to none is because jemalloc can mix objects from different arenas inside of that cache. So we wouldn't be able to guarantee isolation of different providers.

Reply:
We will use one provider and one arena for now, so removed.

Reply:
Done

Reply:
@bratpiorka Done